control-plane/pkg/reconciler/source/v2/sourcev2.go

```diff
@@ -0,0 +1,137 @@
+	internalslst "knative.dev/eventing-kafka-broker/control-plane/pkg/client/internals/kafka/listers/eventing/v1alpha1"
+)
+
+const (
+	DefaultDeliveryOrder = internals.Ordered
+)
+
+type ReconcilerV2 struct {
+	*base.Reconciler
+	ConsumerGroupLister internalslst.ConsumerGroupLister
+	InternalsClient     internalsclient.Interface
+}
+
+func (r *ReconcilerV2) ReconcileKind(ctx context.Context, ks *sources.KafkaSource) reconciler.Event {
+
+	//todo initialize consumer group offsets here?
+
+	consgroup, err := r.reconcileConsumerGroup(ctx, ks)
+	if err != nil {
+		ks.GetConditionSet().Manage(&ks.Status).MarkFalse(consumergroup.KafkaConditionConsumerGroupNotReady, "failed to reconcile consumer group", err.Error())
+		return err
+	}
+	ks.GetConditionSet().Manage(&ks.Status).MarkTrue(consumergroup.KafkaConditionConsumerGroupReady)
+
+	ks.Status.MarkSink(consgroup.Status.SubscriberURI)
+	ks.Status.Placeable = consgroup.Status.Placeable
+	ks.Status.Consumers = *consgroup.Spec.Replicas
+
+	return nil
+}
+
+func (r ReconcilerV2) reconcileConsumerGroup(ctx context.Context, ks *sources.KafkaSource) (*internalscg.ConsumerGroup, error) {
+
+	newcg := &internalscg.ConsumerGroup{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      ks.Spec.ConsumerGroup,
+			Namespace: ks.Namespace,
+			OwnerReferences: []metav1.OwnerReference{
+				*kmeta.NewControllerRef(ks),
+			},
+		},
+		Spec: internalscg.ConsumerGroupSpec{
+			Template: internalscg.ConsumerTemplateSpec{
+				Spec: internalscg.ConsumerSpec{
+					Topics: ks.Spec.Topics,
+					Configs: internalscg.ConsumerConfigs{Configs: map[string]string{
+						"group.id":          ks.Spec.ConsumerGroup,
+						"bootstrap.servers": strings.Join(ks.Spec.BootstrapServers, ","),
+					}},
+					Auth: &internalscg.Auth{
+						NetSpec: &ks.Spec.KafkaAuthSpec.Net,
+					},
+					Delivery: &internalscg.DeliverySpec{
+						Ordering: DefaultDeliveryOrder,
+					},
+					Subscriber: ks.Spec.Sink,
+					CloudEventOverrides: &duckv1.CloudEventOverrides{
+						Extensions: ks.Spec.CloudEventOverrides.Extensions,
+					},
+				},
+			},
+		},
+	}
+
+	cg, err := r.ConsumerGroupLister.ConsumerGroups(ks.GetNamespace()).Get(ks.Spec.ConsumerGroup) //Get by consumer group id
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+
+	if err := newcg.Validate(ctx); err != nil {
+		return nil, fmt.Errorf("failed to validate expected consumer group %s/%s: %w", newcg.GetNamespace(), newcg.GetName(), err)
+	}
+
+	if apierrors.IsNotFound(err) {
+		cg, err := r.InternalsClient.InternalV1alpha1().ConsumerGroups(newcg.GetNamespace()).Create(ctx, newcg, metav1.CreateOptions{})
+		if err != nil && !apierrors.IsAlreadyExists(err) {
+			return nil, fmt.Errorf("failed to create consumer group %s/%s: %w", newcg.GetNamespace(), newcg.GetName(), err)
+		}
+		if apierrors.IsAlreadyExists(err) {
+			return newcg, nil
+		}
+		return cg, nil
+	}
```

Comment:
If the `ks.Spec.ConsumerGroup` is not immutable on the KafkaSource CR this might leak ConsumerGroup resources since 
- a user creates a KafkaSource with `Spec.ConsumerGroup: c1`
- we create a `ConsumerGroup` called `c1`
- a user updates the KafkaSource with `Spec.ConsumerGroup: c2`
- c1 will keep being in the system even though it shouldn't be.

What about using the UUID as name?

Reply:
Ah yes, right. `c1` will still have `ks` as owner which means it wont be deleted until `ks` is deleted too